internal/storage/loopback/loopback.go

```diff
@@ -0,0 +1,191 @@
+	_, span := trace.StartSpan(ctx, "loopback.Mount")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+
+	span.AddAttributes(
+		trace.StringAttribute("target", target),
+		trace.Int64Attribute("device", int64(device)),
+		trace.StringAttribute("backingFile", backingFile))
+
+	if err := osMkdirAll(target, 0700); err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			osRemoveAll(target)
+		}
+	}()
+
+	// If the device already exists `getLoopID` will return -1, so skip this if the device
+	// we're asking for already exists. The UVM starts with 8 loop devices (0,1,2....7)
+	// already so this should be hit exactly 8 times before we have to actually allocate
+	// a new loop device.
+	loopID := device
+	if _, err := os.Stat(fmt.Sprintf("/dev/loop%d", device)); err != nil {
+		loopCtrl, err := getLoopController(context.Background())
+		if err != nil {
+			return errors.Wrap(err, "failed to get loop controller")
+		}
+
+		// Provision loop device
+		loopID = getLoopID(ctx, loopCtrl, device)
+		if loopID == -1 {
+			return errors.New("failed to provision loop device")
+		}
+	}
+
+	// Pair loop dev with backing file
+	source, err := pairLoopDevice(ctx, loopID, backingFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to pair backing file %s to loop device with ID %d", backingFile, loopID)
+	}
+
+	return unixMount(source, target, "ext4", uintptr(unix.MS_RDONLY), "noload")
+}
+
+// Teardown a file backed loop device
+func Teardown(ctx context.Context, device int) error {
+	if err := RemoveBackingFile(ctx, device); err != nil {
+		return errors.Wrap(err, "could not teardown loop device")
+	}
+
+	loopCtrl, err := getLoopController(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get loop controller")
+	}
+
+	// remove loop device
+	deallocateLoopDevice(ctx, loopCtrl, device)
+	return nil
+}
+
+func getLoopController(ctx context.Context) (int, error) {
+	// Get loop controller
+	fd, err := safeOpen(ctx, "/dev/loop-control", unix.O_RDWR, EmptyMode)
+	if err != nil {
+		return 0, err
+	}
+	return fd, nil
+}
+
+func getLoopID(ctx context.Context, loopControllerFD, device int) int {
+	loopID, err := allocateLoopDevice(ctx, loopControllerFD, device)
+	if err != nil {
+		return -1
+	}
+	return loopID
+}
+
+// Implement some methods from https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=770fe30a46a12b6fb6b63fbe1737654d28e8484
+/*
+Note: devnr means device number
+Example in C:
+ cfd = open("/dev/loop-control", O_RDWR);
+
+ # add a new specific loop device
+ err = ioctl(cfd, LOOP_CTL_ADD, devnr);
+
+ # remove a specific loop device
+ err = ioctl(cfd, LOOP_CTL_REMOVE, devnr);
+
+ # find or allocate a free loop device to use
+ devnr = ioctl(cfd, LOOP_CTL_GET_FREE);
+
+ sprintf(loopname, "/dev/loop%i", devnr);
+ ffd = open("backing-file", O_RDWR);
+ lfd = open(loopname, O_RDWR);
+ err = ioctl(lfd, LOOP_SET_FD, ffd);
+*/
```

Comment:
why?

Reply:
This implements some functionality similar to what losetup would do under the hood. Another approach would be to mknod the device nodes but we still need the ioctls to add/remove the backing files

Reply:
The hope is we don't ever actually have to remove the device nodes and simply just remove the backing file and re-use the same node for another container